File: mobile_app/lib/data/user_prefs.dart
import 'package:shared_preferences/shared_preferences.dart';

class UserPrefs {
  static const _languageKey = 'selected_language';
  static const _avatarIdKey = 'selected_avatar_id';
  static const _avatarNameKey = 'selected_avatar_name';
  static const _avatarImageKey = 'selected_avatar_image';
  static const _avatarTeachingStyleKey = 'selected_avatar_teaching_style';
  static const _avatarPersonalityKey = 'selected_avatar_personality';
  static const _avatarVoiceIdKey = 'selected_avatar_voice_id';
  static const _avatarSpeechRateKey = 'selected_avatar_speech_rate';
  static const _avatarPitchLevelKey = 'selected_avatar_pitch_level';
  static const _accessTokenKey = 'auth_access_token';
  static const _userJsonKey = 'auth_user_json';
  static const _familySetupKey = 'family_setup_complete';
  static const _learningFocusKey = 'parent_learning_focus';

  Future<void> saveLanguage(String language) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_languageKey, language);
  }

  Future<String> getLanguage() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_languageKey) ?? 'amharic';
  }

  Future<void> saveAvatar({
    required String id,
    required String name,
    required String imageUrl,
    String? teachingStyle,
    String? personalityDescription,
    String? voiceId,
    double? speechRate,
    double? pitchLevel,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_avatarIdKey, id);
    await prefs.setString(_avatarNameKey, name);
    await prefs.setString(_avatarImageKey, imageUrl);
    if (teachingStyle != null) {
      await prefs.setString(_avatarTeachingStyleKey, teachingStyle);
    }
    if (personalityDescription != null) {
      await prefs.setString(_avatarPersonalityKey, personalityDescription);
    }
    if (voiceId != null) {
      await prefs.setString(_avatarVoiceIdKey, voiceId);
    }
    if (speechRate != null) {
      await prefs.setDouble(_avatarSpeechRateKey, speechRate);
    }
    if (pitchLevel != null) {
      await prefs.setDouble(_avatarPitchLevelKey, pitchLevel);
    }
  }

  Future<String?> getAvatarId() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_avatarIdKey);
  }

  Future<String?> getAvatarName() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_avatarNameKey);
  }

  Future<String?> getAvatarImage() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_avatarImageKey);
  }

  Future<String?> getAvatarTeachingStyle() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_avatarTeachingStyleKey);
  }

  Future<String?> getAvatarPersonality() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_avatarPersonalityKey);
  }

  Future<String?> getAvatarVoiceId() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_avatarVoiceIdKey);
  }

  Future<double?> getAvatarSpeechRate() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getDouble(_avatarSpeechRateKey);
  }

  Future<double?> getAvatarPitchLevel() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getDouble(_avatarPitchLevelKey);
  }

  Future<void> saveAuth({
    required String accessToken,
    required String userJson,
  }) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_accessTokenKey, accessToken);
    await prefs.setString(_userJsonKey, userJson);
  }

  Future<void> markFamilySetupComplete() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setBool(_familySetupKey, true);
  }

  Future<bool> isFamilySetupComplete() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getBool(_familySetupKey) ?? false;
  }

  Future<void> saveLearningFocus(String focus) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString(_learningFocusKey, focus);
  }

  Future<String> getLearningFocus() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_learningFocusKey) ?? 'Reading';
  }

  Future<String?> getAccessToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_accessTokenKey);
  }

  Future<String?> getUserJson() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString(_userJsonKey);
  }

  Future<void> clearAuth() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove(_accessTokenKey);
    await prefs.remove(_userJsonKey);
  }
}

File: mobile_app/src/data/userPrefs.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

const KEYS = {
  language: 'selected_language',
  avatarId: 'selected_avatar_id',
  avatarName: 'selected_avatar_name',
  avatarImage: 'selected_avatar_image',
  avatarTeachingStyle: 'selected_avatar_teaching_style',
  avatarPersonality: 'selected_avatar_personality',
  avatarVoiceId: 'selected_avatar_voice_id',
  avatarSpeechRate: 'selected_avatar_speech_rate',
  avatarPitchLevel: 'selected_avatar_pitch_level',
  accessToken: 'auth_access_token',
  userJson: 'auth_user_json',
  familySetup: 'family_setup_complete',
  learningFocus: 'parent_learning_focus',
};

export const userPrefs = {
  async saveLanguage(language: string) {
    await AsyncStorage.setItem(KEYS.language, language);
  },

  async getLanguage() {
    return (await AsyncStorage.getItem(KEYS.language)) ?? 'amharic';
  },

  async saveAvatar(input: {
    id: string;
    name: string;
    imageUrl: string;
    teachingStyle?: string | null;
    personalityDescription?: string | null;
    voiceId?: string | null;
    speechRate?: number | null;
    pitchLevel?: number | null;
  }) {
    await AsyncStorage.multiSet([
      [KEYS.avatarId, input.id],
      [KEYS.avatarName, input.name],
      [KEYS.avatarImage, input.imageUrl],
    ]);
    if (input.teachingStyle != null) {
      await AsyncStorage.setItem(KEYS.avatarTeachingStyle, input.teachingStyle);
    }
    if (input.personalityDescription != null) {
      await AsyncStorage.setItem(KEYS.avatarPersonality, input.personalityDescription);
    }
    if (input.voiceId != null) {
      await AsyncStorage.setItem(KEYS.avatarVoiceId, input.voiceId);
    }
    if (input.speechRate != null) {
      await AsyncStorage.setItem(KEYS.avatarSpeechRate, String(input.speechRate));
    }
    if (input.pitchLevel != null) {
      await AsyncStorage.setItem(KEYS.avatarPitchLevel, String(input.pitchLevel));
    }
  },

  getAvatarId: () => AsyncStorage.getItem(KEYS.avatarId),
  getAvatarName: () => AsyncStorage.getItem(KEYS.avatarName),
  getAvatarImage: () => AsyncStorage.getItem(KEYS.avatarImage),
  getAvatarTeachingStyle: () => AsyncStorage.getItem(KEYS.avatarTeachingStyle),
  getAvatarPersonality: () => AsyncStorage.getItem(KEYS.avatarPersonality),
  getAvatarVoiceId: () => AsyncStorage.getItem(KEYS.avatarVoiceId),

  async getAvatarSpeechRate() {
    const value = await AsyncStorage.getItem(KEYS.avatarSpeechRate);
    return value == null ? null : Number(value);
  },

  async getAvatarPitchLevel() {
    const value = await AsyncStorage.getItem(KEYS.avatarPitchLevel);
    return value == null ? null : Number(value);
  },

  async saveAuth(accessToken: string, userJson: string) {
    await AsyncStorage.multiSet([
      [KEYS.accessToken, accessToken],
      [KEYS.userJson, userJson],
    ]);
  },

  async markFamilySetupComplete() {
    await AsyncStorage.setItem(KEYS.familySetup, 'true');
  },

  async isFamilySetupComplete() {
    return (await AsyncStorage.getItem(KEYS.familySetup)) === 'true';
  },

  async saveLearningFocus(focus: string) {
    await AsyncStorage.setItem(KEYS.learningFocus, focus);
  },

  async getLearningFocus() {
    return (await AsyncStorage.getItem(KEYS.learningFocus)) ?? 'Reading';
  },

  getAccessToken: () => AsyncStorage.getItem(KEYS.accessToken),
  getUserJson: () => AsyncStorage.getItem(KEYS.userJson),

  async clearAuth() {
    await AsyncStorage.multiRemove([KEYS.accessToken, KEYS.userJson]);
  },
};
